ls1intum/Artemis
# `Grading`: Allow tutors to grade presentations

### Checklist
#### General
#### Server

#### Client

### Motivation and Context
The grading system in use lacks a comprehensive approach to grading presentations. Therefor, this PR aims to add graded presentations as an alternative to the current "presentation score".

### Description
This PR adds a new feature to our grading system that allows instructors to activate graded presentations on the grading key page. Instructors can choose between None and Graded presentations, and if Graded is selected, the grading key page now displays the number of presentations and the combined weight of the presentations. This data is then saved to the grading_scale table in the database. Furthermore, Tutors are now able to assess presentations and enter presentation scores on the participation pages of exercises.

Assignments in prioritised order:
🟡 The instructor can specify in the Course Grading Key that presentations are graded for Course Scores.
-- ✅ If enabled, what % of the grading does this account for? (e.g. 20%)
-- ✅ If enabled, how many presentations should be graded? (e.g. 2)
-- ✅ In the process, change the old mechanism to "presentations" instead of "presentation score".
-- ✅ (DB change in GradingScale)
---- ✅ Double for the share of presentations in the total grade (e.g. 20%) --> nullable
---- ✅ Int for the number of presentations --> nullable
-- ✅ In the User Interface: Option to activate in the Grading Key, if active, these two values must be entered.
-- ✅ One of the two values zero means that the graded presentations are not active.
-- ❌ Follow-up: move the previous presentation score in the course to the GradingScale dialogue. [TODO]
🟡 The tutor can indicate in the Participation Table which score a student had during the presentation.
-- ❌ Follow-Up: there is an error message when entering a possible third presentation. [TODO]
-- ✅ (DB change, if possible simply change participation.presentation_score from int to double)

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Tutor
- 1 Student

 (As an Instructor)
1. Log in to Artemis
2. Navigate to Course Management
3. Navigate to the Grading Key of a Course without enabled presentations (might be enabled in course edit page!)
4. Activate Graded Presentations and insert a valid Grading Key (make sure you cannot save nonsense values for numbers of presentations and weight)
5. Go to an exercise and click on the presentation checkbox

(As a Student)
1. Participate in the exercise that is eligible for presentations

(As a Tutor)
1. Navigate to the participation page of the exercise the student participated in
2. Test adding grades [valid from 0 to 100], changing them and removing them

Verify that tutors do not see buttons that link to pages they cannot access
Verify that other links still work as well

### Review Progress

#### Performance Review

#### Code Review

#### Manual Tests

### Test Coverage



| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|-----------------------------:|
| GradingScale.java | 99% | ✅                           |
| StudentParticipation.java | 97% | ✅              |
| GradingScaleService | 91% | ✅              |
| CourseResource.java | 94% | ✅              |
| ParticipationResource.java | 90% | ✅              |
| - | - | -              |
| presentation-score.component.ts | 88.88% | ✅              |
| base-grading-system.component.ts | 95.03% |  ✅              |
| grading-system-presentations.component.ts | 96.96% | ✅              |


### Screenshots


(Presentation mode selection on Grading Key Page)
<img width="1716" alt="Bildschirmfoto 2023-04-24 um 23 22 54" src="https://user-images.githubusercontent.com/47261058/234119969-f374aa61-577b-4fbc-a307-e977bc11253e.png">

(Participations and option to add grades as a tutor)
<img width="1716" alt="Bildschirmfoto 2023-04-24 um 23 22 20" src="https://user-images.githubusercontent.com/47261058/234119976-f5c63d11-411a-45d7-a495-b1ecf032ade2.png">
